elixir-shyryan/lib/lexer.ex

```diff
@@ -1,77 +1,143 @@
-      <<">", rest::binary>> -> tokenize(rest, [Token.new(:greater_than) | tokens])
-      <<"<", rest::binary>> -> tokenize(rest, [Token.new(:less_than) | tokens])
-      <<"return", rest::binary>> -> tokenize(rest, [Token.new(:return) | tokens])
-      <<_::8, rest::binary>> -> tokenize(rest, [Token.new(:illegal) | tokens])
-    end
+  # Ignore whitespace.
+  defp lex(<<c::8, rest::binary>>, tokens) when is_whitespace(c) do
+    lex(rest, tokens)
   end
 
-  defp read_identifier(input, tokens) do
-    {identifier, rest} = accumulate_identifier(input, "")
+  # Tail-recursively go through the input, tokenizing the character(s).
+  defp lex(input, tokens) do
+    {token, rest} = tokenize(input)
+    lex(rest, [token | tokens])
+  end
+
+  # Uses binary pattern-matching to match on the first character(s).
+  # See: https://hexdocs.pm/elixir/Kernel.SpecialForms.html#<<>>/1
+  @spec tokenize(String.t()) :: {token(), String.t()}
+  defp tokenize(<<c::8, _::binary>> = input) when is_letter(c), do: read_identifier(input)
+  defp tokenize(<<c::8, _::binary>> = input) when is_digit(c), do: read_number(input)
+  defp tokenize(<<"==", rest::binary>>), do: {:equal_equal, rest}
+  defp tokenize(<<"!=", rest::binary>>), do: {:not_equal, rest}
+  defp tokenize(<<";", rest::binary>>), do: {:semicolon, rest}
+  defp tokenize(<<",", rest::binary>>), do: {:comma, rest}
+  defp tokenize(<<"(", rest::binary>>), do: {:lparen, rest}
+  defp tokenize(<<")", rest::binary>>), do: {:rparen, rest}
+  defp tokenize(<<"{", rest::binary>>), do: {:lsquirly, rest}
+  defp tokenize(<<"}", rest::binary>>), do: {:rsquirly, rest}
+  defp tokenize(<<"=", rest::binary>>), do: {:assign, rest}
+  defp tokenize(<<"+", rest::binary>>), do: {:plus, rest}
+  defp tokenize(<<"-", rest::binary>>), do: {:minus, rest}
+  defp tokenize(<<"!", rest::binary>>), do: {:bang, rest}
+  defp tokenize(<<"/", rest::binary>>), do: {:slash, rest}
+  defp tokenize(<<"*", rest::binary>>), do: {:asterisk, rest}
+  defp tokenize(<<">", rest::binary>>), do: {:greater_than, rest}
+  defp tokenize(<<"<", rest::binary>>), do: {:less_than, rest}
+  defp tokenize(<<"return", rest::binary>>), do: {:return, rest}
+  defp tokenize(<<c::8, rest::binary>>), do: {{:illegal, <<c>>}, rest}
 
-    token =
-      identifier
-      |> Token.lookup_identifier()
-      |> Token.new()
+  # Recusrively read the input until we hit a non-letter character. Then
+  # tokenize the word.
+  @spec read_identifier(String.t(), String.t()) :: {token(), String.t()}
+  defp read_identifier(input, acc \\ "")
 
-    tokenize(rest, [token | tokens])
+  defp read_identifier(<<c::8, rest::binary>>, acc) when is_letter(c) do
+    read_identifier(rest, acc <> <<c>>)
   end
 
-  defp accumulate_identifier(<<c::8, rest::binary>>, acc) when is_letter(c) do
-    accumulate_identifier(rest, acc <> <<c>>)
+  defp read_identifier(rest, identifier) do
+    {tokenize_word(identifier), rest}
   end
```

Comment:
Turn the iodata value into a binary

```suggestion
  defp read_identifier(rest, acc) do
    identifier = 
      acc
      |> IO.iodata_to_binary()
      |> tokenize_word()
    
    {identifier, rest}
  end
```